gtsfm/common/view_frustum.py

```diff
@@ -0,0 +1,146 @@
+
+from typing import Tuple
+
+import numpy as np
+from gtsam import Point3, Pose3
+
+DEFAULT_FRUSTUM_RAY_LEN = 0.3  # meters, arbitrary
+
+
+class ViewFrustum:
+    """ Generates edges of a 5-face mesh for drawing pinhole camera in 3d"""
+
+    def __init__(self, fx: float, img_w: int, img_h: int, frustum_ray_len: int = DEFAULT_FRUSTUM_RAY_LEN) -> None:
+        """
+        Args:
+            fx: focal length in x-direction, assuming square pixels (fx == fy)
+            img_w: image width (in pixels)
+            img_h: image height (in pixels)
+            frustum_ray_len: extent to which extend frustum rays away from optical center
+                (increase length for large-scale scenes to make frustums visible)
+        """
+        self.fx_ = fx
+        self.img_w_ = img_w
+        self.img_h_ = img_h
+        self.frustum_ray_len_ = frustum_ray_len
+
+    def get_frustum_vertices(self) -> Tuple[np.ndarray,np.ndarray,np.ndarray,np.ndarray,np.ndarray]:
+        """Obtain 3d positions of all 5 frustum vertices
+
+          (x,y,z)               (x,y,z)                (x,y,z)              (x,y,z)
+              \\=================//                      \\                   //
+               \\               //                        \\ 1-------------2 //
+        (-w/2,-h/2,fx)       (w/2,-h/2,fx)                 \\| IMAGE PLANE |//
+                 1-------------2                             |             |/
+                 |\\         //| IMAGE PLANE  (-w/2, h/2,fx) 3-------------4 (w/2, h/2,fx)
+                 | \\       // | IMAGE PLANE                  \\         //
+                 3--\\-----//--4                               \\       //
+                     \\   //                                    \\     //
+                      \\ //                                      \\   //
+                        O PINHOLE                                 \\ //
+                                                                    O PINHOLE
+        """
+        uv = np.array(
+            [
+                [self.img_w_ // 2, self.img_h_ // 2],  # optical center
+                [0, 0],  # top-left
+                [self.img_w_ - 1, 0],  # top-right
+                [0, self.img_h_ - 1],  # bottom-left
+                [self.img_w_ - 1, self.img_h_ - 1],  # bottom-right
+            ]
+        )
```

Comment:
Doesnt this disagree with the image above?

Reply:
I understood after reading `compute_pixel_ray_directions_vectorized` how it works. But someone reading this method alone will likely be confused.

Reply:
Maybe that function can be modified to accept offsets?